Track cross-tab WindowProxy accesses

bnham opened the pull request:
#### 52d668d5cf530e5a5e5b06ad0535655a4cc58878
<pre>
Track cross-tab WindowProxy accesses
<a href="https://bugs.webkit.org/show_bug.cgi?id=265822">https://bugs.webkit.org/show_bug.cgi?id=265822</a>
<a href="https://rdar.apple.com/118208200">rdar://118208200</a>

Reviewed by Chris Dumez.

We want to track how WindowProxy is used across tabs (i.e. via an opener relationship). This can be
used to build up telemetry on how useful something like `Cross-Origin-Opener-Policy:
restrict-properties` would be on the web (which is a version of COOP that restricts cross-origin
WindowProxy to just the postMessage and closed properties).

To do this, we trap all property access in JSWindowProxy. In my own local tests, this seems to work.
Most WindowProxy property access seems to be just from callers trying to access some property on
their own window object. That bails out early on in the logic added by this patch (in the
`if target == lexicalGlobalObject` check).

From there, if we detect that WindowProxy is used across an opener relationship, we message the
UIProcess about the property access the first time it happens via LocalFrame and
LocalFrameLoaderClient.

Note that the current approach doesn&apos;t handle the case of detecting if a WindowProxy is accessed
through multiple levels of openers (e.g. `window.opener.opener.postMessage`). But I don&apos;t think that
use case is very common.

One additional change is that I had to add some URL loading history to FrameLoadState. This is
because if an opened frame accesses WindowProxy in some way, we want to tell UIProcess not only
about the currently loaded domain in that frame, but also previous domains that have been loaded in
that frame. The URL history is only enabled for main frames (since those might have been opened by a
DOM opener), so I don&apos;t think it should have much memory impact. BackForwardList isn&apos;t sufficient
since we also want to capture things like server-side redirects that occur before the load event
fires in the page.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/opener/child-access-parent-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/child-access-parent-via-windowproxy.html: Added.
* LayoutTests/http/wpt/opener/iframe-access-top-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/iframe-access-top-via-windowproxy.html: Added.
* LayoutTests/http/wpt/opener/parent-access-child-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/parent-access-child-via-windowproxy.html: Added.
* LayoutTests/http/wpt/opener/resources/empty.html: Added.
* LayoutTests/http/wpt/opener/resources/send-message-to-opener.html: Added.
* LayoutTests/http/wpt/opener/resources/send-message-to-top.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/bindings/js/JSWindowProxy.cpp:
(WebCore::frameInfo):
(WebCore::hasSameMainFrame):
(WebCore::logCrossTabPropertyAccess):
(WebCore::checkCrossTabWindowProxyUsage):
(WebCore::JSWindowProxy::getOwnPropertySlot):
(WebCore::JSWindowProxy::getOwnPropertySlotByIndex):
(WebCore::JSWindowProxy::put):
(WebCore::JSWindowProxy::putByIndex):
(WebCore::JSWindowProxy::deleteProperty):
(WebCore::JSWindowProxy::deletePropertyByIndex):
(WebCore::JSWindowProxy::defineOwnProperty):
* Source/WebCore/bindings/js/JSWindowProxy.h:
* Source/WebCore/loader/FrameLoaderTypes.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::setDocument):
(WebCore::LocalFrame::didAccessWindowProxyPropertyViaOpener):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Location.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/FrameLoadState.cpp:
(WebKit::FrameLoadState::didStartProvisionalLoad):
(WebKit::FrameLoadState::didSuspend):
(WebKit::FrameLoadState::didExplicitOpen):
(WebKit::FrameLoadState::didReceiveServerRedirectForProvisionalLoad):
(WebKit::FrameLoadState::didFailProvisionalLoad):
(WebKit::FrameLoadState::didCommitLoad):
(WebKit::FrameLoadState::didFinishLoad):
(WebKit::FrameLoadState::didSameDocumentNotification):
(WebKit::FrameLoadState::setURL):
* Source/WebKit/UIProcess/FrameLoadState.h:
(WebKit::FrameLoadState::Observer::didReceiveProvisionalURL):
(WebKit::FrameLoadState::Observer::didCancelProvisionalLoad):
(WebKit::FrameLoadState::Observer::didCommitProvisionalLoad):
(WebKit::FrameLoadState::Observer::didFinishLoad):
(WebKit::FrameLoadState::setURL): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCreateMainFrame):
(WebKit::WebPageProxy::didAccessWindowProxyPropertyViaOpenerForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::didAccessWindowProxyProperty):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didAccessWindowProxyPropertyViaOpener):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Tools/MiniBrowser/mac/AppDelegate.m:
(windowProxyPropertyDescription):
(-[BrowserAppDelegate websiteDataStore:domain:didOpenDomainViaWindowOpen:withProperty:directly:]):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::getAndClearReportedWindowProxyAccessDomains):
(WTR::TestRunner::didGetAndClearReportedWindowProxyAccessDomains):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::getAndClearReportedWindowProxyAccessDomains):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):
(WTR::createWKArray):
(WTR::TestController::getAndClearReportedWindowProxyAccessDomains):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate init]):
(-[TestWebsiteDataStoreDelegate websiteDataStore:domain:didOpenDomainViaWindowOpen:withProperty:directly:]):
(-[TestWebsiteDataStoreDelegate reportedWindowProxyAccessDomains]):
(-[TestWebsiteDataStoreDelegate clearReportedWindowProxyAccessDomains]):

Canonical link: <a href="https://commits.webkit.org/271791@main">https://commits.webkit.org/271791@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4f0fff48b31e3c19c67bbd76e1062d43c84304d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29680 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8333 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30976 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32177 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30296 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10476 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5607 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29953 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6949 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25310 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5930 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/29833 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/6072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26396 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33517 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/25477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27045 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26819 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/32283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/29886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/6021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/4218 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/30059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7762 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/26195 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/36245 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7041 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7812 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6548 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+    if (!frame)
+        return std::nullopt;
+
+    return FrameInfo { *frame, frame->mainFrame() };
```
Does the caller gracefully handle if the frame goes away?

bnham replied:
I wouldn't think that is possible since I'm not calling back out to JS or something that could cause a deallocation from underneath me. But I don't really know what else can cause a Frame to disappear beneath me and I've been told we're pretty much banned from using raw refs and pointers except for trivial use cases, so I'll switch these to use Refs instead.

johnwilander commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1261,6 +1262,20 @@ void LocalFrame::frameWasDisconnectedFromOwner() const
+    if (!parentWindow)
+        return;
+
+    auto parentOrigin = SecurityOriginData::fromURL(parentWindow->location().url());
```
I always get worried about opaque origins in when doing these kind of things. Just mentioning it so that you can take it into consideration.

bnham replied:
Good point. I'll skip null and opaque origins here

johnwilander commented on Source/WebKit/UIProcess/FrameLoadState.cpp:
```diff
@@ -113,7 +134,17 @@ void FrameLoadState::didFailLoad()
+
+void FrameLoadState::setURL(const URL& url)
+{
+    m_url = url;
```
Was m_url ever set somewhere else before than in didSameDocumentNotification(). If not, we will now make it possible to set m_url to null which was not possible earlier. That's because you left the null check in the caller. We should make sure that's right.

johnwilander commented on Source/WebKit/UIProcess/FrameLoadState.h:
```diff
@@ -64,7 +66,7 @@ class FrameLoadState {
 
     State state() const { return m_state; }
     const URL& url() const { return m_url; }
-    void setURL(const URL& url) { m_url = url; }
```
Aha! I got my answer here. :)

johnwilander commented on Tools/MiniBrowser/mac/AppDelegate.m:
```diff
@@ -497,4 +497,25 @@ - (IBAction)clearDefaultStoreWebsiteData:(id)sender
+    case WKWindowProxyPropertyPostMessage:
+        propertyDescription = "postMessage";
+        break;
+    default:
```
I remember a policy to not have default cases. But maybe that's just in C++? I've been used to having an ASSERT_NOT_REACHED instead.

johnwilander reviewed (commented):
LGTM but I'd like someone else to give it r+ or r-. Ideally someone with Frame loading expertise.

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
     return &static_cast<JSVMClientData*>(vm.clientData)->windowProxySpace();
 }
 
+struct FrameInfo {
```
Since this is a temporary experiment that we'll want to get rid of ASAP, I think we should put all new code behind a build time flag to make the experiment-specific code obvious. It will make our life easier when we want to disable / get rid of the experiment.

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
 }
 
+struct FrameInfo {
+    Frame& frame;
```
We don't allow raw C++ references / data members like this. Please use smart pointers.

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+    Frame& mainFrame;
+};
+
+static std::optional<FrameInfo> getFrameInfo(JSGlobalObject* globalObject)
```
no `get` prefixes for getters in WebKit.

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+    if (!domGlobalObject)
+        return std::nullopt;
+
+    auto* document = dynamicDowncast<Document>(domGlobalObject->scriptExecutionContext());
```
RefPtr

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+    if (!document)
+        return std::nullopt;
+
+    auto* frame = document->frame();
```
RefPtr

bnham replied:
I'll use protectedFrame

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+    if (!childFrame.opener())
+        return;
+
+    auto* parentWindow = childFrame.opener()->window();
```
RefPtr

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+        return;
+
+    auto* parentWindow = childFrame.opener()->window();
+    auto* childWindow = childFrame.window();
```
RefPtr

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+        return;
+
+    String propertyNameDescription;
+    if (std::holds_alternative<PropertyName>(propertyName))
```
Can we please use WTF::switchOn(). It is safer.

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+    WindowProxyProperty property = WindowProxyProperty::Other;
+    auto& builtinNames = WebCore::builtinNames(lexicalGlobalObject->vm());
+
+    if (std::holds_alternative<PropertyName>(propertyName)) {
```
WTF::switchOn()

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+        if (hasSameMainFrame(child.frame.opener(), parent)) {
+            logCrossTabPropertyAccess(child.frame, propertyName);
+
+            if (is<LocalFrame>(child.frame))
```
Please use dynamicDowncast<>() in all new code.

cdumez commented on Source/WebCore/bindings/js/JSWindowProxy.cpp:
```diff
@@ -181,4 +184,164 @@ JSC::GCClient::IsoSubspace* JSWindowProxy::subspaceForImpl(JSC::VM& vm)
+        if (hasSameMainFrame(child.mainFrame.opener(), parent)) {
+            logCrossTabPropertyAccess(child.mainFrame, propertyName);
+
+            if (is<LocalFrame>(child.mainFrame))
```
dynamicDowncast<>()

cdumez commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1261,6 +1262,20 @@ void LocalFrame::frameWasDisconnectedFromOwner() const
+        return;
+    m_accessedWindowProxyPropertiesViaOpener.add(property);
+
+    auto* parentWindow = opener() ? opener()->window() : nullptr;
```
RefPtr

cdumez commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -945,6 +945,7 @@ def headers_for_type(type):
         'WebCore::WheelScrollGestureState': ['<WebCore/PlatformWheelEvent.h>'],
         'WebCore::WillContinueLoading': ['<WebCore/FrameLoaderTypes.h>'],
         'WebCore::WillInternallyHandleFailure': ['<WebCore/FrameLoaderTypes.h>'],
+        'WebCore::WindowProxyProperty': ['<WebCore/FrameLoaderTypes.h>'],
```
Do we really need this? Can't we use `header: <WebCore/FrameLoaderTypes.h>`
and `[CustomHeader]` in the serialization.in instead?

bnham replied:
I tried that but the codegen is still trying to insert this line into `WebPageProxyMessageReceiver.cpp`:

```
#include <WebCore/WindowProxyProperty.h>
```

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -459,6 +460,51 @@ class PageClientProtector {
 };
 
+class WebPageProxyFrameLoadStateObserver : public FrameLoadState::Observer {
+    WTF_MAKE_NONCOPYABLE(WebPageProxyFrameLoadStateObserver); WTF_MAKE_FAST_ALLOCATED;
```
Please put `WTF_MAKE_FAST_ALLOCATED` on the next line.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -459,6 +460,51 @@ class PageClientProtector {
     WeakPtr<PageClient> m_pageClient;
 };
 
+class WebPageProxyFrameLoadStateObserver : public FrameLoadState::Observer {
```
Can the class be marked as final?

bnham committed:
Track cross-tab WindowProxy accesses
https://bugs.webkit.org/show_bug.cgi?id=265822
rdar://118208200

Reviewed by Chris Dumez.

We want to track how WindowProxy is used across tabs (i.e. via an opener relationship). This can be
used to build up telemetry on how useful something like `Cross-Origin-Opener-Policy:
restrict-properties` would be on the web (which is a version of COOP that restricts cross-origin
WindowProxy to just the postMessage and closed properties).

To do this, we trap all property access in JSWindowProxy. In my own local tests, this seems to work.
Most WindowProxy property access seems to be just from callers trying to access some property on
their own window object. That bails out early on in the logic added by this patch (in the
`if target == lexicalGlobalObject` check).

From there, if we detect that WindowProxy is used across an opener relationship, we message the
UIProcess about the property access the first time it happens via LocalFrame and
LocalFrameLoaderClient.

Note that the current approach doesn't handle the case of detecting if a WindowProxy is accessed
through multiple levels of openers (e.g. `window.opener.opener.postMessage`). But I don't think that
use case is very common.

One additional change is that I had to add some URL loading history to FrameLoadState. This is
because if an opened frame accesses WindowProxy in some way, we want to tell UIProcess not only
about the currently loaded domain in that frame, but also previous domains that have been loaded in
that frame. The URL history is only enabled for main frames (since those might have been opened by a
DOM opener), so I don't think it should have much memory impact. BackForwardList isn't sufficient
since we also want to capture things like server-side redirects that occur before the load event
fires in the page.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/opener/child-access-parent-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/child-access-parent-via-windowproxy.html: Added.
* LayoutTests/http/wpt/opener/iframe-access-top-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/iframe-access-top-via-windowproxy.html: Added.
* LayoutTests/http/wpt/opener/parent-access-child-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/parent-access-child-via-windowproxy.html: Added.
* LayoutTests/http/wpt/opener/resources/empty.html: Added.
* LayoutTests/http/wpt/opener/resources/send-message-to-opener.html: Added.
* LayoutTests/http/wpt/opener/resources/send-message-to-top.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/bindings/js/JSWindowProxy.cpp:
(WebCore::frameInfo):
(WebCore::hasSameMainFrame):
(WebCore::logCrossTabPropertyAccess):
(WebCore::checkCrossTabWindowProxyUsage):
(WebCore::JSWindowProxy::getOwnPropertySlot):
(WebCore::JSWindowProxy::getOwnPropertySlotByIndex):
(WebCore::JSWindowProxy::put):
(WebCore::JSWindowProxy::putByIndex):
(WebCore::JSWindowProxy::deleteProperty):
(WebCore::JSWindowProxy::deletePropertyByIndex):
(WebCore::JSWindowProxy::defineOwnProperty):
* Source/WebCore/bindings/js/JSWindowProxy.h:
* Source/WebCore/loader/FrameLoaderTypes.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::setDocument):
(WebCore::LocalFrame::didAccessWindowProxyPropertyViaOpener):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/Location.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
* Source/WebKit/UIProcess/FrameLoadState.cpp:
(WebKit::FrameLoadState::didStartProvisionalLoad):
(WebKit::FrameLoadState::didSuspend):
(WebKit::FrameLoadState::didExplicitOpen):
(WebKit::FrameLoadState::didReceiveServerRedirectForProvisionalLoad):
(WebKit::FrameLoadState::didFailProvisionalLoad):
(WebKit::FrameLoadState::didCommitLoad):
(WebKit::FrameLoadState::didFinishLoad):
(WebKit::FrameLoadState::didSameDocumentNotification):
(WebKit::FrameLoadState::setURL):
* Source/WebKit/UIProcess/FrameLoadState.h:
(WebKit::FrameLoadState::Observer::didReceiveProvisionalURL):
(WebKit::FrameLoadState::Observer::didCancelProvisionalLoad):
(WebKit::FrameLoadState::Observer::didCommitProvisionalLoad):
(WebKit::FrameLoadState::Observer::didFinishLoad):
(WebKit::FrameLoadState::setURL): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCreateMainFrame):
(WebKit::WebPageProxy::didAccessWindowProxyPropertyViaOpenerForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::didAccessWindowProxyProperty):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didAccessWindowProxyPropertyViaOpener):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.h:
* Tools/MiniBrowser/mac/AppDelegate.m:
(windowProxyPropertyDescription):
(-[BrowserAppDelegate websiteDataStore:domain:didOpenDomainViaWindowOpen:withProperty:directly:]):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::didReceiveMessageToPage):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::getAndClearReportedWindowProxyAccessDomains):
(WTR::TestRunner::didGetAndClearReportedWindowProxyAccessDomains):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::getAndClearReportedWindowProxyAccessDomains):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::cocoaResetStateToConsistentValues):
(WTR::createWKArray):
(WTR::TestController::getAndClearReportedWindowProxyAccessDomains):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate init]):
(-[TestWebsiteDataStoreDelegate websiteDataStore:domain:didOpenDomainViaWindowOpen:withProperty:directly:]):
(-[TestWebsiteDataStoreDelegate reportedWindowProxyAccessDomains]):
(-[TestWebsiteDataStoreDelegate clearReportedWindowProxyAccessDomains]):

Canonical link: https://commits.webkit.org/271791@main
